### PR TITLE
Fix save restore of frontstage layout

### LINKED ISSUE #842: User's widget layout preferences not saved
### Describe the bug

**Expected behavior**
Users have the ability to move widgets around to redefined their UI layout as they want. When a user leaves the viewer, he's expecting that his layout will be saved throughout his sessions with the viewer.

**Current behavior**
The user's layout isn't saved properly.

### To Reproduce

1. Checkout and run apps/test-viewer from https://github.com/iTwin/viewer-components-react with any model
2. Drag any widget to another section of any panel. (https://github.com/iTwin/appui/assets/147163568/7ef37697-bee5-4a4c-8579-d76ab6447d13)
3. Reload the page
4. Notice the widget layout is reset

** 2.5 - Reorder widgets within the same section. (https://github.com/iTwin/appui/assets/147163568/327a1b09-2036-41b2-8c3a-761b081a64e9)

### Expected Behavior

Dragging any widget to another section or reordering widgets tabs should be preserved throughout users' sessions.

### Screenshots

Initial load local storage key: 
image

After the user has moved a widget to another section: 
image

After reordering widget tabs: 
image

After reloading: 
image

### Desktop (please complete the applicable information)

Chrome Version 124.0.6367.119 (Official Build) (64-bit)

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by GerardasB**
## Changes

This PR fixes #842 caused when React18 is used with `StrictMode`.
React and testing library versions were updated.

## Testing

Added additional unit tests.

### CODE REVIEW

File: ui/appui-react/src/test/backstage/BackstageComposer.test.tsx [hunk 1]
```diff
@@ -256,7 +255,7 @@ describe("BackstageComposer", () => {
     act(() => UiItemsManager.unregister(uiProvider.id));
   });
 
-  describe("useGroupedItems", () => {
+  describe.only("useGroupedItems", () => {
```
Comment (ignas-k) [inline ui/appui-react/src/test/backstage/BackstageComposer.test.tsx:258]: Forgot to remove

File: ui/appui-react/src/test/cursor/cursorpopup/CursorPopup.test.tsx [hunk 2]
```diff
@@ -242,7 +242,8 @@ describe("CursorPopup", () => {
 
   // After looking thoroughly the numbers, the "working" tests are wrong
   // This needs to be completely reviewed...
-  it("should set offset if more than one popup in a position", async () => {
+  // TODO: react 18 upgrade
```
Comment (ignas-k) [inline ui/appui-react/src/test/cursor/cursorpopup/CursorPopup.test.tsx:245]: Just out of curiosity: what do these mean? Are we waiting for something?

**Review by ignas-k — APPROVED** [review]

File: ui/appui-react/src/test/cursor/cursorpopup/CursorPopup.test.tsx (hunk 2 above)
Comment (GerardasB) [reply ui/appui-react/src/test/cursor/cursorpopup/CursorPopup.test.tsx:245]: I've skipped some of the tests that started failing randomly. Especially with the comment above.

File: ui/appui-react/src/test/backstage/BackstageComposer.test.tsx (hunk 1 above)
Comment (GerardasB) [reply ui/appui-react/src/test/backstage/BackstageComposer.test.tsx]: Fixed.

**Comment by GerardasB** [comment]
@Mergifyio backport release/4.13.x

**Pull request merged**